### PR TITLE
internal: add global DialOptions and ServerOptions for all clients and servers

### PR DESCRIPTION
**Pull request by lidizheng**
This PR adds internal only methods to set default DialOption and ServerOption. Note that, it's setting the mutators not the actual dialOptions or serverOptions struct.

```go
import "google.golang.org/grpc/internal"

internal.SetDefaultServerOption(StatsHandler(A), StatsHandler(B))
internal.SetDefaultDialOption(WithStatsHandler(A), WithStatsHandler(B))
```

RELEASE NOTES: none

### CODE REVIEW

**Comment by lidizheng** [comment]
@dfawley @menghanl PTAL.

**Comment by lidizheng** [comment]
Bump.

**Comment by lidizheng** [comment]
Bump again. PTAL.

**Comment by dfawley** [comment]
Sorry for the delay here.  Can you update to resolve the conflicts please?

File: dialoptions.go [hunk 1]
```diff
@@ -35,6 +35,12 @@ import (
 	"google.golang.org/grpc/stats"
 )
 
+func init() {
+	internal.SetDefaultDialOption = func(opt ...DialOption) {
+		defaultDialOption = opt
```
Comment (dfawley) [inline dialoptions.go:40]: `defaultDialOption = append(defaultDialOption, opt...)` & rename to `AddDefaultDialOptions`?

File: dialoptions.go [hunk 2]
```diff
@@ -70,6 +76,8 @@ type DialOption interface {
 	apply(*dialOptions)
 }
 
+var defaultDialOption []DialOption
```
Comment (dfawley) [inline dialoptions.go:79]: Nit: `defaultDialOptions`

**Review by dfawley — COMMENTED** [review]
This LGTM.  However, can you please add a test for these features (server and client)?

**Review by lidizheng — COMMENTED** [review]
Test added, about the naming of the internal functions, let's continue our discussion in the thread. PTALA.

File: dialoptions.go (hunk 1 above)
Comment (dfawley) [reply dialoptions.go:40]: For tests we can add a `Set` (or a `Clear`, which seems better for cleanup).  There's a pretty big usability problem with `Set` for normal operation, since it only allows for a single user.

File: dialoptions.go (hunk 2 above)
Comment (dfawley) [reply dialoptions.go:79]: Let's not use the word "default" here, then, this is too confusing.  How about "extra" or "additional" or "injected" or "plugin"instead?  Or if you have any better ideas, I'm happy to hear them.

File: dialoptions.go [hunk 3]
```diff
@@ -76,7 +79,7 @@ type DialOption interface {
 	apply(*dialOptions)
 }
 
-var defaultDialOption []DialOption
+var extraDefaultDialOption []DialOption
```
Comment (dfawley) [inline dialoptions.go:82]: "extraDefault"? How about just "extra"?

File: dialoptions.go [hunk 4]
```diff
@@ -36,8 +36,11 @@ import (
 )
 
 func init() {
-	internal.SetDefaultDialOption = func(opt ...DialOption) {
-		defaultDialOption = opt
+	internal.AddDefaultDialOptions = func(opt ...DialOption) {
```
Comment (dfawley) [inline dialoptions.go:39]: `Extra` here too please.

**Comment by lidizheng** [comment]
The failure is unrelated.

File: dialoptions.go [hunk 5]
```diff
@@ -557,7 +568,7 @@ func withHealthCheckFunc(f internal.HealthChecker) DialOption {
 	})
 }
 
-func defaultDialOptions() dialOptions {
+func newDefaultDialOptions() dialOptions {
```
Comment (dfawley) [inline dialoptions.go:571]: Why the change here?

**Review by dfawley — APPROVED** [review]

File: dialoptions.go (hunk 5 above)
Comment (lidizheng) [reply dialoptions.go]: Reverted.

You mentioned it's a bit confusing around the naming, and update the name helps to differentiate the variable and the factory method.
